ad/views.py

```diff
@@ -99,6 +101,55 @@ def get_queryset(self):
         model = self.get_model()
         return model.objects.all()
     
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        ad = self.get_object()
+        adtype = self.kwargs['adtype']
+
+        if ad.owner != self.request.user:
+            context['comments'] = Message.objects.filter(
+                content_type__model=adtype+'ad',
+                object_id=ad.id,
+                user=self.request.user
+            )
+        else:
+            context['comments'] = Message.objects.filter(
+                content_type__model=adtype+'ad',
+                object_id=ad.id
+            ).exclude(user=self.request.user)
+
+        context['adtype'] = adtype
+        context['form'] = MessageForm()
+
+        return context
+    
+    def post(self, request, *args, **kwargs):
+        ad = self.get_object()
+        adtype = self.kwargs['adtype']
+        pk = self.kwargs['pk']
+        form = MessageForm(request.POST)
+        
+        if form.is_valid():
+            new_comment = form.save(commit=False)
+            new_comment.ad = ad
+            new_comment.user = request.user
+            
+            parent_id = request.POST.get('parent_id')
```

Comment:
Instead of passing parent comment id in URL, allow the MessageForm itself to expect parent field. You can add a hidden form field in HTML for parent.